### DISCUSSION TITLE
Batch insert using DAO

### DISCUSSION THREAD

**Discussion by dacianb** [pocketbase/pocketbase#1763] [Q&A] [answered]
Hello everyone,

After some research on the topic, i found mixed results...
I currently work on a personal side project... and i want to batch insert/update records from a public API.
Since there is no such functionality out of the box i implemented something crude using the RunInTransaction() method. Is there any way better then this to do this kind of task?
Other then creating a raw SQL expression.

Thanks!

```go
app.Dao().RunInTransaction(func(txDao *daos.Dao) error {
		collection, err := txDao.FindCollectionByNameOrId("brands")
		if err != nil {
			return err
		}

		rows := make([]dbx.NullStringMap, len(brands)+1)

		for key, value := range brands {
			rows[key] = map[string]sql.NullString{
				"name":    {String: value.Name, Valid: true},
				"gsma_id": {String: strconv.Itoa(value.Id), Valid: true},
			}
		}

		record := models.NewRecordsFromNullStringMaps(collection, rows)

		for _, rec := range record {
			rec.RefreshId()
			rec.MarkAsNew()
			rec.RefreshCreated()
			rec.RefreshUpdated()
			if err := txDao.SaveRecord(rec); err != nil {
				return err
			}
		}
		return nil
	})
```

**Comment by ganigeorgiev** [comment, marked as answer]
`RunInTransaction` is the correct way to insert group of records.

You don't need to explicitly call `RefreshId`, `RefreshCreated`, `RefreshUpdated` since they will be called automatically on save. Since the records don't have a custom id `MarkAsNew` is also not necessary in this case. Here is an slightly shorter version of the above:

```go
app.Dao().RunInTransaction(func(txDao *daos.Dao) error {
        // optionally, you can move the FindCollectionByNameOrId out of the transaction
        // to minimize eventual SQLITE_BUSY errors on high concurrent load
        // due to the mixed read and write transactions locking
        // (https://www.sqlite.org/lang_transaction.html#read_transactions_versus_write_transactions)
	collection, err := txDao.FindCollectionByNameOrId("brands")
	if err != nil {
		return err
	}

	for _, brand := range brands {
		record := models.NewRecord(collection)
		record.Set("name", brand.Name)
		record.Set("gsma_id", brand.Id) // you don't necessary need to explicitly cast the field value
		                                // it will be automatically casted to string if the gsma_id is `text` field

		if err := txDao.SaveRecord(record); err != nil {
			return err
		}
	}

	return nil
})
```

**Reply by dacianb** [reply]
@ganigeorgiev 
Thanks a lot! Your version is much more slick and efficient.
Using my version i needed to call the `RefreshId()` etc. methods, otherwise i would get a SQL error with `ID not set` and the created and updated columns would be empty. Also if i did not call `MarkAsNew()` the query would be a Update not Insert statement.

Anyway you should a example like this to the docs since it's a really common type of operation imho.

**Reply by ganigeorgiev** [reply]
> Using my version i needed to call the RefreshId() etc. methods, otherwise i would get a SQL error with ID not set and the created and updated columns would be empty.

Yes, it is possible because the `NewRecordsFromNullStringMaps` and `NewRecordFromNullStringMap` methods are intended to load and scan data from the db and will call `record.PostScan()`, aka. the record will be marked as "not new" which on save will trigger the `update` scenario instead of `create`.

I've added it in my todo and will consider documenting it better in the near future.

**Comment by Sleepful** [comment]
Just leaving the note here, since [v23.0.0-rc release:](https://github.com/pocketbase/pocketbase/releases/tag/v0.23.0-rc) 

> Batch/transactional Web API endpoint.

**Reply by azliR** [reply]
Hi, a quick question, with current transactional web API, is it possible to do like
```dart
final record = pb.collection(name).create(field);
final secondRecord = pb.collection(name2).create({"id": record.id, ...});
```
Or is it just a batch insert?
